File: CashuWallet/Models/CashuRequest.swift
import Foundation

struct CashuRequest: Codable, Identifiable, Hashable {
    let id: String
    let encoded: String
    let amount: UInt64?
    let unit: String
    let mints: [String]
    let memo: String?
    let createdAt: Date
    var receivedPaymentIds: [String]

    init(
        id: String = Self.newId(),
        encoded: String,
        amount: UInt64? = nil,
        unit: String = "sat",
        mints: [String] = [],
        memo: String? = nil,
        createdAt: Date = Date(),
        receivedPaymentIds: [String] = []
    ) {
        self.id = id
        self.encoded = encoded
        self.amount = amount
        self.unit = unit
        self.mints = mints
        self.memo = memo
        self.createdAt = createdAt
        self.receivedPaymentIds = receivedPaymentIds
    }

    static func newId() -> String {
        UUID().uuidString.split(separator: "-").first.map(String.init) ?? UUID().uuidString
    }
}

File: android/app/src/main/java/org/cashu/wallet/Models/Requests/CashuRequest.kt
package org.cashu.wallet.Models

import java.util.UUID
import kotlinx.serialization.Serializable

@Serializable
data class CashuRequestPayment(
    val transactionId: String,
    val amount: Long,
    val receivedAtEpochMillis: Long,
)

@Serializable
data class CashuRequest(
    val id: String = newId(),
    val encoded: String,
    val amount: Long? = null,
    val unit: String = "sat",
    val mints: List<String> = emptyList(),
    val memo: String? = null,
    val createdAtEpochMillis: Long = System.currentTimeMillis(),
    val receivedPayments: List<CashuRequestPayment> = emptyList(),
    val receivedPaymentIds: List<String> = emptyList(),
) {
    val totalReceived: Long get() = receivedPayments.sumOf { it.amount }

    fun withLegacyPaymentFallback(): CashuRequest {
        if (receivedPayments.isNotEmpty() || receivedPaymentIds.isEmpty()) return this
        return copy(
            receivedPayments = receivedPaymentIds.map { id ->
                CashuRequestPayment(
                    transactionId = id,
                    amount = 0,
                    receivedAtEpochMillis = createdAtEpochMillis,
                )
            },
            receivedPaymentIds = emptyList(),
        )
    }

    companion object {
        fun newId(): String = UUID.randomUUID().toString().substringBefore("-")
    }
}
